specs/networking/rpc-interface.md

```diff
@@ -0,0 +1,257 @@
+ETH 2.0 Networking Spec - RPC Interface
+===
+
+# Abstract
+
+The Ethereum 2.0 networking stack uses two modes of communication: a broadcast protocol that gossips information to interested parties via GossipSub, and an RPC protocol that retrieves information from specific clients. This specification defines the RPC protocol.
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL", NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this document are to be interpreted as described in RFC 2119.
+
+# Dependencies
+
+This specification assumes familiarity with the [Messaging](./messaging.md), [Node Identification](./node-identification), and [Beacon Chain](../core/0_beacon-chain.md) specifications.
+
+# Specification
+
+## Message Schemas
+
+Message body schemas are notated like this:
+
+```
+(
+    field_name_1: type
+    field_name_2: type
+)
+```
+
+Embedded types are serialized as SSZ Containers unless otherwise noted.
+
+All referenced data structures can be found in the [0-beacon-chain](https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#data-structures) specification.
+
+## `libp2p` Protocol Names
+
+A "Protocol ID" in `libp2p` parlance refers to a human-readable identifier `libp2p` uses in order to identify sub-protocols and stream messages of different types over the same connection. Peers exchange supported protocol IDs via the `Identify` protocol upon connection. When opening a new stream, peers pin a particular protocol ID to it, and the stream remains contextualised thereafter. Since messages are sent inside a stream, they do not need to bear the protocol ID.
+
+## RPC-Over-`libp2p`
+
+To facilitate RPC-over-`libp2p`, a single protocol path is used: `/eth/serenity/beacon/rpc/1.0.0`. Remote method calls are wrapped in a "request" structure:
```

Comment:
Just curious about the reason why we have only one protocol id, instead of "one protocol for every type of request". If we have only one protocol id and use SSZ as the codec, from my understanding we might need a `method_id` prefix(either as a varint, or a fixed byte like in RLPx) for each message. This way we can read all of the bytes for a message, and then decode them according to their `method_id`

Reply:
`libp2p` incurs stream switching overhead every time the protocol changes. We do have to extract the method ID here, however this can be done with byte offsets in SSZ.

Reply:
Let me correct an imprecision there. There’s no such thing as protocol changes inside a stream. When you open a stream, you pin a protocol to it and it stays invariant. So if you have N protocols and want to exchange messages for all of them, you open at least N streams.

Reply:
Am I misunderstanding the `multistream-select` spec? From my interpretation of https://github.com/multiformats/multistream-select, for one TCP connection to send/receive messages for a given protocol a new stream needs to be selected by sending a 'switch protocols' message. Is this accurate?

Reply:
While my nomenclature may be incorrect (apologies if it is), I'm referring to any overhead incurred on-the-wire to make two different clients speak the same protocol.

Reply:
It's true that at the moment, libp2p will write the textual identifier of the protocol at the start of each opened stream and this can be considered "stream switching overhead" when making requests. The problem will be addressed in MultiStream 2.0 though, so I think we can design the wire protocol with this upgrade in mind:

https://github.com/libp2p/specs/pull/95

With Multistream 2.0, there will be an initial handshake step determining the shared protocols between two peers. Then all shared protocols will be assigned numeric IDs which will be used when opening new streams.

Reply:
In other words, the ETH clients will be much simpler if `RPC-Over-libp2p` is defined to just create a new stream for each requests. We don't need to duplicate the multiplexing capabilities of libp2p in our own code.

Reply:
I can't tell if there is a timeline for Multistream 2.0 to be released. We can always update this spec to use Multistream 2.0 when the spec is ready, but for now I'm in favor of using the stable `libp2p` to avoid introducing implied dependencies.